nucypher/crypto/keystore.py

```diff
@@ -0,0 +1,434 @@
+        return True
+
+    @staticmethod
+    def __save(secret: bytes, password: str, keystore_dir: Optional[Path] = None) -> Path:
+        failures = validate_keystore_password(password)
+        if failures:
+            # TODO: Ensure this scope is separable from the scope containing the password
+            #       to help avoid unintentional logging of the password.
+            raise InvalidPassword(''.join(failures))
+
+        # Derive verifying key (used as ID)
+        verifying_key = _derive_umbral_key(material=secret, info=_VERIFYING_INFO)
+        keystore_id = verifying_key.to_bytes().hex()[:Keystore._ID_SIZE]
+
+        # Generate paths
+        keystore_dir = keystore_dir or Keystore._DEFAULT_DIR
+        os.makedirs(abspath(keystore_dir), exist_ok=True, mode=0o700)
+        keystore_path = generate_keystore_filepath(parent=keystore_dir, id=keystore_id)
+
+        # Encrypt secret
+        __salt = token_bytes(_SALT_SIZE)
+        __wrapping_key = _derive_wrapping_key(salt=__salt, password=password)
+        encrypted_secret = bytes(SecretBox(__wrapping_key).encrypt(secret))
+
+        # Create keystore file
+        keystore_payload = _assemble_keystore(encrypted_secret=encrypted_secret, salt=__salt)
+        _write_keystore(path=keystore_path, payload=keystore_payload, serializer=_serialize_keystore)
+
+        return keystore_path
+
+    #
+    # Public API
+    #
+
+    @classmethod
+    def load(cls, id: str, keystore_dir: Path = _DEFAULT_DIR) -> 'Keystore':
+        filepath = generate_keystore_filepath(parent=keystore_dir, id=id)
+        instance = cls(keystore_path=filepath)
+        return instance
+
+    @classmethod
+    def restore(cls, words: str, password: str, keystore_dir: Optional[Path] = None) -> 'Keystore':
+        """Restore a keystore from seed words"""
+        __mnemonic = Mnemonic(_MNEMONIC_LANGUAGE)
+        __secret = __mnemonic.to_entropy(words)
+        path = Keystore.__save(secret=__secret, password=password, keystore_dir=keystore_dir)
+        keystore = cls(keystore_path=path)
+        return keystore
+
+    @classmethod
+    def generate(cls, password: str, keystore_dir: Optional[Path] = None, interactive: bool = True) -> 'Keystore':
+        """Generate a new nucypher keystore for use with characters"""
+        mnemonic = Mnemonic(_MNEMONIC_LANGUAGE)
+        __words = mnemonic.generate(strength=_ENTROPY_BITS)
+        if interactive:
+            cls._confirm_generate(__words)
+        __secret = mnemonic.to_entropy(__words)
+        path = Keystore.__save(secret=__secret, password=password, keystore_dir=keystore_dir)
+        keystore = cls(keystore_path=path)
+        return keystore
+
+    @staticmethod
+    def _confirm_generate(__words: str) -> None:
+        """
+        Inform the caller of new keystore seed words generation the console
+        and optionally perform interactive confirmation.
+        """
+
+        # notification
+        emitter = StdoutEmitter()
```

Comment:
Can't say I like `StdoutEmitter()` popping up here and there in the codebase...

Reply:
Me neither.  I even opened an Issue about it once (#1754).  Alternatives are either to pass it down from the top (or at least a level higher) or use `click` outright, but this will bypass the logic of emitters and like logging controls.

Reply:
I would be in favor of reopening that issue.